src/rust-2024/rpit-lifetime-capture.md

```diff
@@ -1,12 +1,313 @@
-# RPIT lifetime capture
+# RPIT lifetime capture rules
 
 🚧 The 2024 Edition has not yet been released and hence this section is still "under construction".
 
-This feature is partially implemented, and not yet ready for testing.
-More information may be found in the tracking issue at <https://github.com/rust-lang/rust/issues/117587>.
+This chapter describes changes related to the **Lifetime Capture Rules 2024** introduced in [RFC 3498], including opaque type *precise capturing* introduced in [RFC 3617].
+
+[RFC 3498]: https://github.com/rust-lang/rfcs/pull/3498
+[RFC 3617]: https://github.com/rust-lang/rfcs/pull/3617
 
 ## Summary
 
+- In all editions, `use<..>` bounds allow for capturing only the lifetime parameters needed in RPIT opaque types.
+- If a `use<..>` bound is elided (i.e. not present) then the compiler automatically inserts one according to edition-specific rules.
+- In Rust 2024, *all* in-scope generic parameters, including lifetime parameters, are included in these automatically inserted `use<..>` bounds.
+- Uses of the `Captures` trick (`Captures<..>` bounds) and of the outlives trick (e.g. `'_` bounds) can be replaced by `use<..>` bounds (in all editions) or removed entirely (in Rust 2024).
+
 ## Details
 
+### Capturing and `use<..>` syntax
+
+*Capturing* a generic parameter in an RPIT (return-position impl Trait) opaque type allows for that parameter to be used in the corresponding hidden type.  In Rust 1.82, we added `use<..>` bounds that allow specifying explicitly which generic parameters to capture.  E.g.:
+
+```rust
+# #![feature(precise_capturing)]
+fn capture<'a, T>(x: &'a (), y: T) -> impl Sized + use<'a, T> {
+    //                                ~~~~~~~~~~~~~~~~~~~~~~~
+    //                             This is the RPIT opaque type.
+    //
+    //                                It captures `'a` and `T`.
+    (x, y)
+  //~~~~~~
+  // The hidden type is: `(&'a (), T)`.
+  //
+  // This type can use `'a` and `T` because they were captured.
+}
+```
+
+Currently, `use<..>` bounds must list all generic type and const parameters in scope, but they may exclude in-scope lifetime parameters.  E.g.:
+
+```rust
+# #![feature(precise_capturing)]
+fn capture<'a, T>(_: &'a (), _: T) -> impl Sized + use<T> {
+    //                                ~~~~~~~~~~~~~~~~~~~
+    //                 The opaque type captures only `T` and not `'a`.
+    ()
+}
+```
+
+You might want to capture more lifetimes in order to return a hidden type that uses those lifetimes.  Conversely, you might want to capture fewer lifetimes in order to allow the opaque type to be used in more places.  E.g., this is an error because the lifetime is captured despite the fact that the hidden type does not use the lifetime:
```

Comment:
> You might want to capture more lifetimes [...]

Would be nice to show an example of how comes about on ed < 2024.

> You might want to capture fewer lifetimes [...]

Would be nice to show an example of how this happens on ed 2024.

Reply:
There are examples later in the document when we start talking about the migration.  Maybe we could go on a bit more there, but this chapter is already rather long, and at some point it feels like people should just read RFC 3498 if they want the full motivation.

Capturing is a subtle enough topic that the approach taken here is to introduce and it explain it first before getting into the edition-specific differences in the next section.

Reply:
This entire paragraph has now been removed from the document.